fw/ss_skb.c

```diff
@@ -1609,3 +1609,289 @@ ss_skb_add_frag(struct sk_buff *skb_head, struct sk_buff *skb, char* addr,
+ * The caller is responsible to keep track of the original skb.
+ * @off - offset within linear data of @orig_skb
+ * @f_pos - position of newly added fragment within @nskb->frags
+ */
+static void
+__skb_linear_to_frag(const struct sk_buff *orig_skb, struct sk_buff *nskb,
+		     unsigned int off, unsigned int f_pos)
+{
+	struct page *data_page;
+	BUG_ON(f_pos >= MAX_SKB_FRAGS);
+
+	data_page = virt_to_page(orig_skb->data);
+	__skb_fill_page_desc(nskb, f_pos, data_page,
+			     orig_skb->data + off -
+			     (unsigned char *)page_to_virt(data_page),
+			     skb_headlen(orig_skb) - off);
+
+	__skb_frag_ref(&skb_shinfo(nskb)->frags[f_pos]);
+	skb_shinfo(nskb)->nr_frags++;
+	ss_skb_adjust_data_len(nskb, skb_headlen(orig_skb) - off);
+}
+
+/* Move @cnt fragment descriptors of the @skb starting from @off
+ * to @nskb starting from @skb_shinfo(@nskb)->frags[@noff]
+ */
+static void
+__skb_frags_transfer(struct sk_buff *skb, struct sk_buff *nskb,
+		     unsigned int noff, unsigned int cnt, unsigned int off)
+{
+	int i;
+	if (!skb_shinfo(skb)->nr_frags)
+		return;
+
+	BUG_ON(noff + cnt > MAX_SKB_FRAGS);
+	BUG_ON(off + cnt > MAX_SKB_FRAGS);
+
+	memcpy(&skb_shinfo(nskb)->frags[noff], &skb_shinfo(skb)->frags[off],
+	       cnt * sizeof(skb_frag_t));
+	skb_shinfo(nskb)->nr_frags += cnt;
+	for (i = noff; i < noff + cnt; i++) {
+		skb_frag_t *f = &skb_shinfo(nskb)->frags[i];
+		ss_skb_adjust_data_len(nskb, skb_frag_size(f));
+		__skb_frag_ref(f);
+	}
+}
+
+/* Using @split_point transform the remaining linear portion of original @skb
+ * to a fragment of the new SKB. Take care of fragments of the @skb too.
+ * If data in the linear area is not needed, just check for available head room,
+ * so the won't be any problems pushing transport headers later.
+ */
+int
+ss_skb_linear_transform(struct sk_buff *skb, struct sk_buff **skb_head,
+			unsigned char *split_point)
+{
+	struct sk_buff *nskb;
+	struct skb_shared_info *si;
+	BUG_ON(!skb_headlen(skb));
+
+	si = skb_shinfo(skb);
+	if (!split_point) {
+		/* Usage of linear portion of SKB is not expected */
+		skb_pull(skb, skb_headlen(skb));
+	} else {
+		nskb = ss_skb_alloc(0);
+		if (!nskb)
+			goto out_err;
+
+		__skb_linear_to_frag(skb, nskb,
+				     split_point - skb->data, 0);
+
+		/* can we move all the original fragments of @skb to @nskb */
+		if (likely(si->nr_frags < MAX_SKB_FRAGS)) {
+			__skb_frags_transfer(skb, nskb, 1, si->nr_frags, 0);
+			ss_skb_queue_tail(skb_head, nskb);
+		} else {
+			/* @skb contains maximum amount of fragments already */
+			ss_skb_queue_tail(skb_head, nskb);
+
+			nskb = ss_skb_alloc(0);
+			if (!nskb) {
+				__kfree_skb(*skb_head);
+				goto out_err;
+			}
+			__skb_frags_transfer(skb, nskb, 0, si->nr_frags, 0);
+			ss_skb_queue_tail(skb_head, nskb);
+		}
+	}
+	return 0;
+
+out_err:
+	T_WARN("%s: can't allocate new skb for response"
+	       " transformation(orig skb %pK)\n",
+	       __func__, skb);
+	return -ENOMEM;
+}
```

Comment:
This looks like a specialized version of `__split_linear_data()`: we should either use `__split_linear_data()` or unify both the functions to make them share the code. It's not good to have two sets of functions doing nearly the same.